src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -176,7 +179,15 @@ Task ITransactionalResource.Abort(long transactionId)
             // Learning that t is committed implies that all pending transactions before t also committed
             if (transactionId > this.stableVersion)
             {
-                bool success = await this.storageExecutor.AddNext(() => GuardState(() => PersistCommit(transactionId)));
+                try
+                {
+                    bool success = await this.storageExecutor.AddNext(() => GuardState(() => PersistCommit(transactionId)));
+                } catch(Exception)
+                {
+                    // On error, queue up a recovery action.  Will do nothing if state recovers before this is processed
+                    this.storageExecutor.AddNext(() => GuardState(() => Task.FromResult(true))).Ignore();
```

Comment:
I don't quite understand the reasoning behind this. Just to clarify:
1. Is this merely an optimization to attempt to refresh the state from storage before the next call begins?
2. Could this be moved into GuardState?

Reply:
Yes, this attempts to reread the state from storage prior to the next persist call (either from prepare or commit).  It does this because we don't know when the next persist call will come through and we don't want to leave the facet in a bad state any longer than is necessary.  On persist error we queue up a recovery action which will try once to recover the state, if it does so prior to the next commit, the next commit will not need to recover first, if it fails, the persist will recover before attempting the persist.